Refactor error messages

hinerm opened the pull request:
This change reduces the number of `helpVerbose` calls during failed matches and shunts the full debugging stack trace of agglomerated exceptions to `log.debug`.

So in Fiji, for example, the script:
```groovy
#@ OpEnvironment ops

ops.op("deconvolve.richardsonLucyTV").input(1, 2, 3, 4, 5).function()
```

Produces the error message:
```
org.scijava.ops.api.OpMatchingException: No Ops found matching this request.
	at org.scijava.ops.engine.impl.DefaultOpEnvironment.findOp(DefaultOpEnvironment.java:397)
	at org.scijava.ops.engine.impl.DefaultOpEnvironment.op(DefaultOpEnvironment.java:212)
	at org.scijava.ops.api.OpBuilder.matchFunctionHelper(OpBuilder.java:21528)
	at org.scijava.ops.api.OpBuilder.matchFunction(OpBuilder.java:21297)
	at org.scijava.ops.api.OpBuilder$Arity5_IV_OU.function(OpBuilder.java:4043)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at New_.run(New_.groovy:3)
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:331)
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:161)
	at org.scijava.plugins.scripting.groovy.GroovyScriptLanguage$1.eval(GroovyScriptLanguage.java:97)
	at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:173)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:165)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:125)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:64)
	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:247)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

but still dumps the full 2k lines to debug

hinerm committed:
Extract No Ops message to constant

hinerm committed:
Move Op suggestion generation to top-level call

Op matches generate a large history of exceptions, as there are multiple
matchers and only one will ultimately succeed (if any). Generating
potential alternatives for the user is a helpful way to point them to
ways to correct their Op call, but doing so for each matching attempt
means we then need to display all the failures to the user, to the point
of being unhelpful. It also means an order of magnitude more Op matches
are attempted for any match request that is destined to fail, as each
failing request goes through a help call to identify near matches.

Instead, we now make this help call once at the complete end of
matching. The agglomerated exceptions from matching are logged at a
debug level but then discarded, leaving a single simple stack trace with
the resulting help text (presumably to display to users).

This may result in a loss of information, as there are no help requests
for any converted parameters or adaptation attempts. If that is
determined to be useful and necessary then it could be built into the
help functionality.

Closes https://github.com/scijava/scijava/issues/184

hinerm commented:
Note that this completely hoses [several exception tests](https://github.com/scijava/incubator/blob/cfd86a133dbc0c841c2adc02b215bdb238e68cfb/scijava-ops-engine/src/test/java/org/scijava/ops/engine/matcher/OpMatchingExceptionTest.java#L78-L144).

It also makes the [DependencyMatchingException](https://github.com/scijava/incubator/blob/cfd86a133dbc0c841c2adc02b215bdb238e68cfb/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DependencyMatchingException.java#L53) subclass not useful.

My feeling is that the `DependencyMatchingException` wasn't really useful anyway given the cluster of error types that appear during matching. This problem is reflected in just how broad the scope of these tests are: "did the error message of interest appear _somewhere_ in this mass of agglomerated exceptions?" In these cases we know these messages indicate the "true" problem, but naively looking at the exception messages would not concretely show that this was the root cause.

That is the whole motivation of this change: that a user is unlikely to know the cause of failure from the exception set alone, so don't overwhelm them, let's just put that info in debugging output. We could perhaps still replicate the behavior of these tests by making a [custom log appender](https://www.baeldung.com/log4j2-custom-appender) that could then be interrogated for the message of choice but it didn't seem trivial to do so and it is something I would file an issue for after paper 1 (or as a paper 1 stretch goal).

@elevans @gselzer @ctrueden I'm going to explore a little more to see if these assumptions fully hold up, but any thoughts/feelings?

hinerm committed:
Further simplify error messages

This change completely removes the individual matcher failures from the
top-level error reporting, instead standardizing to just report the
failed request.

To avoid a total loss of information we special-case some failures
(duplicate Ops).

Note that this does require generalizing some of the exception tests
(although this could be expanded if there was an easy way to capture
debug log output in the tests).

Also removes empty lines between Op descriptions (when present) and
parameters.

hinerm commented:
Update:
I tried compromising and keeping the _messages_ of all the suppressed failures, but still dumping their _stack traces_ to `log.debug`. In general I think things got more manageable when the `helpVerbose` call was moved.

The following are some sample exception scenarios and their messages. It's not perfect but I think better than discarding, or keeping, everything?

1. `missingDependencyViaAdaptationTest`
```
org.scijava.ops.engine.DependencyMatchingException: Could not instantiate dependency of: org.scijava.ops.engine.matcher.MissingDependencyOpArr1|Adaptor:|Info:org.scijava.ops.engine.adapt.functional.ComputersToFunctionsViaFunction$Computer1ToFunction1ViaFunction@0-SNAPSHOT{|Info:org.scijava.ops.engine.create.CreateOpCollection$createDoubleArrayInputAware@0-SNAPSHOT{}}
Dependency identifier: op

Attempted request:
Name: "test.nonexistingOp", Types: java.util.function.Function<java.lang.Double, java.lang.Double>
Input Types: 
		* java.lang.Double
Output Type: 
		* java.lang.Double

	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.resolveOpDependencies(DefaultOpEnvironment.java:650)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.resolveOpDependencies(DefaultOpEnvironment.java:572)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.instantiateOp(DefaultOpEnvironment.java:499)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.generateCacheHit(DefaultOpEnvironment.java:461)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.findOp(DefaultOpEnvironment.java:384)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.op(DefaultOpEnvironment.java:212)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunctionHelper(OpBuilder.java:21528)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunction(OpBuilder.java:21241)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder$Arity1_IV_OT.function(OpBuilder.java:1155)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder$Arity1_IV_OT.apply(OpBuilder.java:1189)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.OpMatchingExceptionTest.missingDependencyViaAdaptationTest(OpMatchingExceptionTest.java:136)
       <collapsed internal unit test lines>
```

2. `ops.op("deconvolve.richardsonLucy").input(1, 2, 3, 4, 5).function()`
```
org.scijava.ops.api.OpMatchingException: No Ops found matching this request.
See debugging output for full failure report.

---Matcher failures---
No matching 'deconvolve.richardsonLucy/org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>' op

Request:
-	Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any

Candidates:
-	 No candidates found!
---
Unable to find an Op adaptation for Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any
---
No matching 'deconvolve.richardsonLucy/org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>' op

Request:
-	Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any

Candidates:
-	 No candidates found!
---


	at org.scijava.ops.engine.impl.DefaultOpEnvironment.findOp(DefaultOpEnvironment.java:415)
	at org.scijava.ops.engine.impl.DefaultOpEnvironment.op(DefaultOpEnvironment.java:212)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunctionHelper(OpBuilder.java:21528)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunction(OpBuilder.java:21297)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder$Arity5_IV_OU.function(OpBuilder.java:4043)
	at org.scijava.ops.image@0-SNAPSHOT/org.scijava.ops.image.deconvolve.DeconvolveTest.fooTest(DeconvolveTest.java:64)
      <collapsed internal unit test lines>
```

3. `duplicateErrorRegressionTest`
```
org.scijava.ops.api.OpMatchingException: No Ops found matching this request.
See debugging output for full failure report.
Perhaps you meant one of these:
test.duplicateOp:
	- org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateA
		> input1 : java.lang.Double
		Returns : java.lang.Double
	- org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateB
		> input1 : java.lang.Double
		Returns : java.lang.Double

---Matcher failures---
Multiple 'test.duplicateOp/java.util.function.Function<java.lang.Double, java.lang.Double>' ops of priority 0.0:
1. org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateA
	> input1 : java.lang.Double
	Returns : java.lang.Double
2. org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateB
	> input1 : java.lang.Double
	Returns : java.lang.Double

Request:
-	Name: "test.duplicateOp", Types: java.util.function.Function<java.lang.Double, java.lang.Double>
Input Types: 
		* java.lang.Double
Output Type: 
		* java.lang.Double

Candidates:
1. 	org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateA
	> input1 : java.lang.Double
	Returns : java.lang.Double
	MATCH
2. 	org.scijava.ops.engine.matcher.OpMatchingExceptionTest$duplicateB
	> input1 : java.lang.Double
	Returns : java.lang.Double
	MATCH

---
Unable to find an Op adaptation for Name: "test.duplicateOp", Types: java.util.function.Function<java.lang.Double, java.lang.Double>
Input Types: 
		* java.lang.Double
Output Type: 
		* java.lang.Double
---
No matching 'test.duplicateOp/java.util.function.Function<java.lang.Double, java.lang.Double>' op

Request:
-	Name: "test.duplicateOp", Types: java.util.function.Function<java.lang.Double, java.lang.Double>
Input Types: 
		* java.lang.Double
Output Type: 
		* java.lang.Double

Candidates:
-	 No candidates found!
---


	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.findOp(DefaultOpEnvironment.java:415)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.impl.DefaultOpEnvironment.op(DefaultOpEnvironment.java:212)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunctionHelper(OpBuilder.java:21528)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunction(OpBuilder.java:21241)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder$Arity1_IT_OT.function(OpBuilder.java:983)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.OpMatchingExceptionTest.duplicateErrorRegressionTest(OpMatchingExceptionTest.java:81)
   <collapsed internal unit test lines>
```

gselzer reviewed (commented):
This is great! I have no suggestions on your changeset!

With that said, one thing that I see now that I find hinders my parsing of the exception is the remaining repetition. For example, in the below example you posted, I see:
* An identical `OpRequest` three different times - what do you think about writing it once at the top of the error message, and then omit it in the matching exceptions? 
* The direct matching routine error message is *identical* to the conversion matching routine error message - we should probably differentiate it somehow.

If you want we could merge this and haggle out my points above in other issues, but might as well resolve them now. Opinions, @elevans @ctrueden?
```
org.scijava.ops.api.OpMatchingException: No Ops found matching this request.
See debugging output for full failure report.

---Matcher failures---
No matching 'deconvolve.richardsonLucy/org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>' op

Request:
-	Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any

Candidates:
-	 No candidates found!
---
Unable to find an Op adaptation for Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any
---
No matching 'deconvolve.richardsonLucy/org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>' op

Request:
-	Name: "deconvolve.richardsonLucy", Types: org.scijava.function.Functions.Arity5<java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, java.lang.Integer, org.scijava.types.Any>
Input Types: 
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
		* java.lang.Integer
Output Type: 
		* org.scijava.types.Any

Candidates:
-	 No candidates found!
---


	at org.scijava.ops.engine.impl.DefaultOpEnvironment.findOp(DefaultOpEnvironment.java:415)
	at org.scijava.ops.engine.impl.DefaultOpEnvironment.op(DefaultOpEnvironment.java:212)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunctionHelper(OpBuilder.java:21528)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder.matchFunction(OpBuilder.java:21297)
	at org.scijava.ops.api@0-SNAPSHOT/org.scijava.ops.api.OpBuilder$Arity5_IV_OU.function(OpBuilder.java:4043)
	at org.scijava.ops.image@0-SNAPSHOT/org.scijava.ops.image.deconvolve.DeconvolveTest.fooTest(DeconvolveTest.java:64)
      <collapsed internal unit test lines>
```

hinerm commented:
> * An identical `OpRequest` three different times - what do you think about writing it once at the top of the error message, and then omit it in the matching exceptions?

I agree this would be ideal. I'll give it a shot.

>The direct matching routine error message is identical to the conversion matching routine error message - we should probably differentiate it somehow.

Also agree here.. I think we just need that matcher to explicitly say it couldn't find any parameter conversion pathways when it fails, like the adaptation matcher does. On it!

hinerm committed:
Differentiate Conversion matching failures

Adds a top-level message recording the fact that this failure is coming
from Conversion matching (as otherwise the exception can appear
identical to Direct matching)

hinerm commented:
@gselzer @elevans OK, I differentiated conversion matching's exception and reconciled the request to just appear at the top-level. Here are some output samples from various failure scenarios. If these seem good enough then let's merge.

```
org.scijava.ops.api.OpMatchingException: Duplicate ops of equal priority detected for request:

Name: "test.duplicateOp", Types: java.util.function.Function<java.lang.Double, java.lang.Double>
Input Types: 
		* java.lang.Double
Output Type: 
		* java.lang.Double


See debugging output for full failure report.
```

```
org.scijava.ops.engine.DependencyMatchingException: Error matching dependencies for request:

Name: "test.adaptMissingDep", Types: java.util.function.Function<[Ljava.lang.Double;, [Ljava.lang.Double;>
Input Types: 
		* java.lang.Double[]
Output Type: 
		* java.lang.Double[]

See debugging output for full failure report.
```

```
org.scijava.ops.api.OpMatchingException: No Ops found matching this request.

Name: "math.add", Types: java.util.function.BiFunction<java.lang.String, java.lang.String, org.scijava.types.Any>
Input Types: 
		* java.lang.String
		* java.lang.String
Output Type: 
		* org.scijava.types.Any

Perhaps you meant one of these:
math.add:
	- org.scijava.ops.engine.math.MathOpCollection$addDoublesOperator
		> input1 : java.lang.Double
		> input2 : java.lang.Double
		Returns : java.lang.Double
	- org.scijava.ops.engine.math.Add$MathPointwiseAddIterablesFunction
		> input1 : I
		> input2 : I
		Returns : java.lang.Iterable<java.lang.Double>
	- org.scijava.ops.image.math.BinaryNumericTypeMath$adder
		 Sum two numeric types
		> input1 : T
		> input2 : T
		> container1 : @CONTAINER T
	- org.scijava.ops.engine.math.Add$MathAddBigIntegersComputer
		> input1 : java.math.BigInteger
		> input2 : java.math.BigInteger
		Returns : java.math.BigInteger
	- org.scijava.ops.engine.math.Add$MathAddDoublesFunction
		> input1 : java.lang.Double
		> input2 : java.lang.Double
		Returns : java.lang.Double
	- org.scijava.ops.engine.math.Add$MathAddNumbersFunction
		> input1 : M
		> input2 : M
		Returns : java.lang.Double
	- org.scijava.ops.engine.math.Add$MathPointwiseAddDoubleArraysComputer
		> input1 : double[]
		> input2 : double[]
		> container1 : @CONTAINER double[]
	- org.scijava.ops.engine.math.Add$MathPointwiseAddDoubleArraysInplace1
		> mutable1 : @MUTABLE double[]
		> input1 : double[]
	- org.scijava.ops.image.math.BinaryRealTypeMath$adder
		 Sets the real component of an output real number to the addition of the
		 real components of two input real numbers.
		> input1 : I1
		> input2 : I2
		> container1 : @CONTAINER O
	- org.scijava.ops.engine.math.MathOpCollection$addDoublesFunction
		> input1 : java.lang.Number
		> input2 : java.lang.Number
		Returns : java.lang.Double

See debugging output for full failure report.
```

gselzer commented:
This looks great to me!
